### ISSUE TITLE
Feature request for country code ACL

### ISSUE THREAD

**Issue by amastelek** [cloudradar-monitoring/rport#670]
This ability as per Openconnect https://www.infradead.org/ocserv/recipes-ocserv-country-blocking.html would be a good feature to have as an ACL for a tunnel.

Example, would be to restrict access to a RDP port to only one country.

**Comment by thorstenkramm** [comment]
@amastelek Interesting. What kind of tunnels do you want to protect with this country whitelisting?
Why is the /32 subnet ACL is not suitable?

**Comment by thorstenkramm** [comment]
@amastelek The quoted service `https://freegeoip.net` redirects to `https://ipstack.com/` which is no longer a free service. You would have to enter your credentials to fetch the country of an IP address. The free plan with just 100 requests per month would not be sufficient.

**Comment by meisanerd** [comment]
You could look at implementing a lookup against the Maxmind GeoIP Lite database, and just have the user provide their own API key for downloading that.  And then do the Country code lookup locally in rport instead of hitting an external service to do it.

In the near-term, this can also possibly be done by setting up firewall rules on the ports on the server hosting rport, instead of in rport itself.

**Comment by thorstenkramm** [comment]
An implementation with a built-in lookup without calling an external API is doable. But I still don't understand the need. Can you please elaborate your use case and which problem you want to solve with the requested feature?

**Comment by amastelek** [comment]
The use is is road warriors requiring RDP access traveling within a specific country.  Mostly various 3g/lte services and hotspots.   Better to have a service restricted to a particular country than to leave it open to the world.  Less risk.

**Comment by thorstenkramm** [comment]
But those road warriors can open the tunnel just for the IP address they have at the moment they want to connect. That achieves maximum security without effort. Why don't you go for this approach?

Resolving IP addresses to countries is not 100% accurate. Especially if you use LTE near the country border. Users might get locked out. But there is no communication channel to explain why and how to solve it.

**Comment by amastelek** [comment]
> But those road warriors can open the tunnel just for the IP address they have at the moment they want to connect. That achieves maximum security without effort. Why don't you go for this approach?
> 
> Resolving IP addresses to countries is not 100% accurate. Especially if you use LTE near the country border. Users might get locked out. But there is no communication channel to explain why and how to solve it.

Yes.  That will work for competent individuals but the extra flow of login, create a tunnel and then launch it each time would need to be reduced to one click from a dashboard.

**Comment by sempervictus** [comment]
GeoIP matching (and deterrent action) can be handled at the OS layer via XTables Addons:
```shell
$ iptables-save 
# Generated by iptables-save v1.8.3 on Sun Aug 28 14:00:52 2022
*filter
:INPUT ACCEPT [347474:26398264]
:FORWARD ACCEPT [0:0]
:OUTPUT ACCEPT [346491:29298817]
-A INPUT -p tcp -m geoip ! --source-country ZZ,US,CA  -j TARPIT --tarpit 
-A INPUT -m geoip ! --source-country ZZ,US,CA  -j DROP
-A OUTPUT -m geoip ! --source-country ZZ,US,CA  -j DROP
COMMIT
# Completed on Sun Aug 28 14:00:52 2022
```

**Comment by bloesss** [comment]
I think I understand the intention as I thought about it too.

Sometimes, our team has to work with the same tunnel at the same time.
With some team members working in homeoffice, the only possibility is to allow the tunnel to any IP.
So at least, with Geo IP Filtering, you could chose to set the ACL to "Germany" for example.

I think that this is something the firewall of your network should do.

We are solving this by putting the tunnel and GUI Ports behind our VPN.
So the tunnels can just be reached over VPN, we can work together from different WAN IPs.

To keep our fqdn, we let the DNS being resolved by the VPN Server, pointing our FQDN to the internal rport server ip.

That way, our security holes are down to the rport connection server that the clients connect to, and our VPN Server.
The rport connection server is behind a reverse proxy that works behind a geo ip filtered firewall, because not all our customers have fixed WAN IP Adresses.

Wouldn't that be the most secure and comfortable solution?
The "any IP" ACL is a very bad idea that never should be needed, because you always open minimum secured ports.

**Comment by thorstenkramm** [comment]
If opening a tunnel without IP ACL is secure or not depends on the protocol used. 
Opening an SSH port worldwide, I wouldn't consider this insecure. Doing it with VNC it's highly dangerous.
RDP also has strong encryption. Security mainly depends on the password strength. 

Experienced users can set up port knocking. All the tunnel ports are blocked by ip tables. With the port knocking client, the user creates a temporary exemption on the firewall. The port knocking demon does it for you. The user only need to fire the port knocking client with the shared secret before using the tunnel.

**Comment by sempervictus** [comment]
@thorstenkramm - respectfully disagree: asym crypto implementations are _hard_ (i am not a formal cryptologist by any means, but i do work with some now and again, and have written a "fair amount" of crypto impl in my time), and there are many functions outside of crypto, including some performed before session init, which are not formally proven (mathematical proof) to be bounded to _only_ doing what they are supposed to without side-effects. I think that the OpenSSL heartbleed problem illustrates this concern rather well in that it was a side-effect of the memory access modality which caused the data leak. Concurrently, please do not leave RDP open to the WAN - SChannel is insanely unsafe and there's a lot more than just crypto problems at play in that stack. BlueKeep and friends are not over, nor will they be until the protocol is rewritten and implemented separately from the legacy WinAPI authentication subsystems making things "convenient" for users and passing around every context of credential you can imagine internally or over nets. A Windows RDP service open on WAN is not a matter of if but a matter of when. Really, any complex service, which performs resource (from hardware to logical) access that is not fully bounded and proofed/proven to be side-effect-free is dangerous to have on the WAN from an (security) architectural perspective.
When going for "world-open" context, i tend to lean toward pre-shared keys used to initiate a cipher state with no negotiation such that the first bits on the wire can't be used by anyone without the keys already in-hand - take a look at how OSSEC handles this for a venerable example (@dcid hit the nail on the head two decades ago). With SSH, or pretty much any negotiated system, there is a fair amount of cleartext exchange to negotiate state (i've written enough of the SSH implementation in Ruby for my Metasploit work to be aware that its not as magically safe as most presume).
If you want to stick to asym crypto in Go, the DARE implementation from MinIO might be worthwhile looking into - they leak a few bytes of the nonce, but its a supported implementation by a commercial entity so piggy-backing off of them might give you some R&D standoff. Personally i think that getting symmetric on PSK or pre-configured-state asym (think 0RTT without a negotiated prior state) would be the safer way to go because that comms stream is the only thing differentiating an RMM tool _you_ control from a RAT of which someone like me takes control... :wink:. There's no silver bullet in security, but if you make barriers to entry so high that they seem insurmountable to the attacker, they will move on to something they _can_ pwn (if they're not an automaton tasked with this one subject domain). Look at it this way: as an attacker, my time is valuable; so do i spend it beating on an opaque fully encrypted service or do i move on to phishing  your users and taking other approaches to realize the value of my time?
